packages/roosterjs-editor-core/lib/coreAPI/insertNode.ts

```diff
@@ -60,29 +64,41 @@ const insertNode: InsertNode = (core: EditorCore, node: Node, option: InsertOpti
 
             // Final check to see if the inserted node is a block. If not block and the ask is to insert on new line,
             // add a DIV wrapping
-            if (insertedNode && insertOnNewLine && !isBlockElement(insertedNode)) {
+            if (insertedNode && option.insertOnNewLine && !isBlockElement(insertedNode)) {
                 wrap(insertedNode);
             }
 
             break;
+        case ContentPosition.Range:
         case ContentPosition.SelectionStart:
+            // Selection start replaces based on the current selection.
+            // Range inserts based on a provided range.
+            // Both have the potential to use the current selection to restore cursor position
+            // So in both cases we need to store the selection state.
             let range = core.api.getSelectionRange(core, true /*tryGetFromCache*/);
-            if (!range) {
-                return;
+            let rangeToRestore = null;
+            if (option.position == ContentPosition.Range) {
+                rangeToRestore = range;
+                range = option.range;
+            } else {
+                if (!range) {
+                    return;
+                } else {
+                    // Create a clone (backup) for the selection first as we may need to restore to it later
+                    rangeToRestore = range.cloneRange();

```

Comment:
Branching logic like this is hard to track. Please pull this out into a private `getInitialRange(options)` or similar